lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -2896,80 +3037,43 @@ struct SplitEmitter : public RootEmitterBase {
   explicit SplitEmitter(StringRef dirname, ModuleOp rootOp)
       : RootEmitterBase(rootOp), dirname(dirname) {}
 
-  /// A list of modules and their position within the per-file operations.
-  struct EmittedModule {
-    Operation *op;
-    size_t position;
-    SmallString<32> filename;
-  };
-  SmallVector<EmittedModule, 0> moduleOps;
-
-  void emitMLIRModule();
-
-private:
   /// The directory to emit files into.
   StringRef dirname;
 
-  /// A list of per-file operations (e.g., `sv.verbatim` or `sv.ifdef`).
-  SmallVector<Operation *, 0> perFileOps;
-
-  void emitModule(const LoweringOptions &options, EmittedModule &mod);
+  void emitMLIRModule();
+  void createFile(const LoweringOptions &options, Identifier fileName,
+                  FileInfo &file);
 };
 
 } // namespace
 
 void SplitEmitter::emitMLIRModule() {
+  gatherFiles(true);
+
   // Load any emitter options from the top-level module.
   LoweringOptions options(rootOp);
 
-  // Partition the MLIR module into modules and interfaces for which we create
-  // separate output files, and the remaining top-level verbatim SV/ifdef
-  // business that needs to go into each file.
-  for (auto &op : *rootOp.getBody()) {
-    TypeSwitch<Operation *>(&op)
-        .Case<HWModuleOp, InterfaceOp>([&](auto &) {
-          moduleOps.push_back({&op, perFileOps.size(), {}});
-        })
-        .Case<VerbatimOp, IfDefProceduralOp, TypeScopeOp>(
-            [&](auto &) { perFileOps.push_back(&op); })
-        .Case<HWGeneratorSchemaOp, HWModuleExternOp>([&](auto &) {})
-        .Default([&](auto *) {
-          op.emitError("unknown operation");
-          encounteredError = true;
-        });
-  }
-
-  // In parallel, emit each module into its separate file, embedded within the
-  // per-file operations.
-  llvm::parallelForEach(moduleOps.begin(), moduleOps.end(),
-                        [&](EmittedModule &mod) { emitModule(options, mod); });
+  // Emit operations to separate files in parallel.
+  llvm::parallelForEach(files.begin(), files.end(), [&](auto &it) {
```

Comment:
Please check `getContext().isMultithreadingEnabled()` and do this in serial if not set.  This allows us to do better A/B testing with multithreading etc.

Reply:
Good point, will update this before merging.

Reply:
I would love to see something like [rayon](https://docs.rs/rayon) in LLVM/MLIR.

Reply:
Agreed, LLVM's parallism support isn't great.